manuals/nsfdbexport.1
.Dd June 25, 2026
.Dt NSFDBEXPORT 1
.Os
.Sh NAME
.Nm nsfdbexport
.Nd exports items stored in a Notes Storage Facility (NSF) database file
.Sh SYNOPSIS
.Nm nsfdbexport
.Op Fl l Ar log_file
.Op Fl t Ar target
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm nsfdbexport
is a utility to export items stored in a Notes Storage Facility (NSF) database \
file
.Pp
.Nm nsfdbexport
is part of the
.Nm libnsfdb
package.
.Nm libnsfdb
is a library to access the Notes Storage Facility (NSF) database file format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl l Ar log_file
logs information about the exported items
.It Fl t Ar target
specify the target directory to export to (default is the source filename \
followed by .export)
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# nsfdbexport bookmark.nsf
nsfdbexport 20260625
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr nsfdbinfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libnsfdb/issues
.Sh COPYRIGHT
Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* complete libnsfdb_io_handle_clear
* format support
  - try reading non-summary notes
* debug output:
  - libnsfdb_io_handle_read_superblock print compression type string
* add dpgk files
* update msvscpp files
* nsfdbinfo:
  - refactor to use info handle
* code clean up
  - remove intptr_t free/clone functions
  - check if already set in initialize
  - use goto
* integrate libfvalue
* superblock
  - non-summary bucket descriptors
* libfdatetime support for nsf timedate ALLDAY (0xffffffffL) definition
  - if time is -1 only date is shown
* change RRV vectors in lists
  - what about the initial value ?

* io handle add functions for
  - unk table lookup

Library
* improve number of notes detection
  - handle 'deleted' notes

API
* get database name
* get other database info like categories, class, design class

Debug
* database flags
  - add flags to definitions
  - add flags to print function
* database replication flags
  - add flags to definitions
  - add flags to print function
* note status flags
  - unknown status flags
* note item field flags
  - add flags to definitions
  - add flags to print function
* print note item type
  - complete libnsfdb_note_item_type.[ch] and implement in libnsfdb_unique_name_key_table.c
* add note item value print function

File format
* bitmaps
* superblock
  - non-summary bucket descriptors
  - modifier note log
  - data note identifier table
  - folder directory object

Tools
* nsfdbinfo
  - print database info
* nsfdbexport
  - export note items
  - add export mode (all, data, non-data)

Deployment
* check local library functions in configure

20141012
* removed README.macosx
* updated dependencies
* worked on utility scripts
* changes for project site move
* see `git log' for more recent change log

20140318
* Code clean up

20140317
* 2014 update

20130922
* 2013 update

20121112
* 2012 update
* updated dependencies
  - libcerror update
  - libcdata update
* code clean up
  - added on_error to nsfdbtools
  - refactored file into export handle
* updated pc file
* updated spec file

20111217
* updated configure.ac and m4 files
* added libfcache
* updated libcstring, libsystem, libuna, libbfio, libfdata, libfdatetime,
  libfvalue
* code clean up

20111030
* updated configure.ac and m4 files
* fixed makefiles for libfvalue
* updated libfdatetime, libfvalue

20111008
* 2011 update
* updated configure.ac and m4 files
* added gettext support
* updated include/error.h, include/types.h
* updated common, libcstring, liberror, libnotify, libsystem
* updated libuna, libbfio, libfdata, libfvalue, libfdatetime
* changes to code for updates
* updated pc and spec file
* updated msvscpp files

20101120
* updated configure.ac, acinclude.m4
* added libnsfdb.rc
* added libfvalue
* updated include/types.h, include/error.h
* updated common, libcstring, liberror, libnotify, libsystem, libuna, libbfio,
  libfdata, libfdatetime
* changes due to libfdata update
* changes due to libsystem update

20100713
* worked on reading nsf files

20100712
* worked on reading nsf files
* changed bucket vector in to list
* worked on nsfexport

20100711
* worked on reading nsf files

20100710
* updated configure.ac and acinclude.m4
* updated common, libcstring, liberror, libnotify, libuna, libfdatetime,
  libbfio, libsystem
* added libfdata
* change amount of in number of
* updated XOR32 with memory alignment support
* update manuals
* created nsfdbexport
* added libfdata support

20100425
* worked on reading nsf files

20100418
* updated libcstring, liberror, libuna, libbfio, libfdatetime

20100408
* worked on reading nsf files

20100407
* worked on reading nsf files

20100406
* worked on reading nsf files

20100405
* worked on reading nsf files

20100403
* worked on reading nsf files

20100402
* renamed project to libnsfdb to prevent name clash with existing libnsf
  for the NES sound file format
* worked on reading nsf files

20100401
* worked on reading nsf files

20100331
* worked on reading nsf files
* updated msvscpp files

20100330
* worked on reading nsf files

20100329
* added libcstring
* updated liberror, libnotify, libuna, libbfio, libsystem, libfdatetime
* worked on reading nsf files
* removed common/narrow_string.h common/wide_string.h

20100322
* worked on msvscpp files
* worked on reading nsf files

20100321
* initial version
* worked on reading nsf files



nsfdbtools/nsfdbexport.c
/*
 * Extracts items from a Notes Storage Facility (NSF) database file
 *
 * Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <memory.h>
#include <narrow_string.h>
#include <system_string.h>
#include <types.h>
#include <wide_string.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "export_handle.h"
#include "log_handle.h"
#include "nsfdbtools_getopt.h"
#include "nsfdbtools_libcerror.h"
#include "nsfdbtools_libclocale.h"
#include "nsfdbtools_libcnotify.h"
#include "nsfdbtools_libcpath.h"
#include "nsfdbtools_libnsfdb.h"
#include "nsfdbtools_output.h"
#include "nsfdbtools_signal.h"
#include "nsfdbtools_unused.h"

export_handle_t *nsfdbexport_export_handle = NULL;
int nsfdbexport_abort                      = 0;

/* Signal handler for nsfdbexport
 */
void nsfdbexport_signal_handler(
      nsfdbtools_signal_t signal NSFDBTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "nsfdbexport_signal_handler";

	NSFDBTOOLS_UNREFERENCED_PARAMETER( signal )

	nsfdbexport_abort = 1;

	if( nsfdbexport_export_handle != NULL )
	{
		if( export_handle_signal_abort(
		     nsfdbexport_export_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal export handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use nsfdbexport to export items stored in a Notes Storage Facility (NSF) database file.";

	nsfdbtools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'l', "log_file", "logs information about the exported items" },
		{ 't', "target", "specify the target directory to export to (default is the source filename followed by .export)" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source file" },
	};
	system_character_t options_string[ 32 ];

	libcerror_error_t *error               = NULL;
	log_handle_t *log_handle               = NULL;
	system_character_t *log_filename       = NULL;
	system_character_t *option_target_path = NULL;
	system_character_t *path_separator     = NULL;
	system_character_t *source             = NULL;
	char *program                          = "nsfdbexport";
	system_integer_t option                = 0;
	size_t source_length                   = 0;
	int number_of_options                  = (int) ( sizeof( options ) / sizeof( nsfdbtools_option_t ) );
	int result                             = 0;
	int verbose                            = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
             "nsfdbtools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( nsfdbtools_output_initialize(
             _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	nsfdbtools_output_version_fprint(
	 stdout,
	 program );

	if( nsfdbtools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = nsfdbtools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				nsfdbtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				nsfdbtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'l':
				log_filename = optarg;

				break;

			case (system_integer_t) 't':
				option_target_path = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				nsfdbtools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		nsfdbtools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	if( option_target_path == NULL )
	{
		source_length = system_string_length(
		                 source );

		path_separator = system_string_search_character_reverse(
		                  source,
		                  (system_character_t) LIBCPATH_SEPARATOR,
		                  source_length );

		if( path_separator == NULL )
		{
			path_separator = source;
		}
		else
		{
			path_separator++;
		}
		option_target_path = path_separator;
	}
	libcnotify_verbose_set(
	 verbose );
	libnsfdb_notify_set_stream(
	 stderr,
	 NULL );
	libnsfdb_notify_set_verbose(
	 verbose );

	if( log_handle_initialize(
	     &log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize log handle.\n" );

		goto on_error;
	}
	if( export_handle_initialize(
	     &nsfdbexport_export_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize export handle.\n" );

		goto on_error;
	}
#if defined( __clang_analyzer__ )
	__builtin_assume( nsfdbexport_export_handle != NULL );
#endif
	if( export_handle_set_target_path(
	     nsfdbexport_export_handle,
	     option_target_path,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set target path.\n" );

		goto on_error;
	}
	result = export_handle_create_items_export_path(
	          nsfdbexport_export_handle,
	          &error );

	if( result == -1 )
	{
		fprintf(
		 stderr,
		 "Unable to create items export path.\n" );

		goto on_error;
	}
	else if( result == 0 )
	{
		fprintf(
		 stderr,
		 "%" PRIs_SYSTEM " already exists.\n",
		 nsfdbexport_export_handle->items_export_path );

		goto on_error;
	}
	if( log_handle_open(
	     log_handle,
	     log_filename,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open log file: %" PRIs_SYSTEM ".\n",
		 log_filename );

		goto on_error;
	}
	fprintf(
	 stdout,
	 "Opening file.\n" );

#ifdef TODO_SIGNAL_ABORT
	if( nsfdbtools_signal_attach(
	     nsfdbexport_signal_handler,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to attach signal handler.\n" );

		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#endif
	if( export_handle_open_input(
	     nsfdbexport_export_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open file: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	if( export_handle_export_file(
	     nsfdbexport_export_handle,
	     log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to export file.\n" );

		goto on_error;
	}
#ifdef TODO_SIGNAL_ABORT
	if( nsfdbtools_signal_detach(
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to detach signal handler.\n" );

		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#endif
	if( export_handle_close_input(
	     nsfdbexport_export_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Error closing file: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	if( export_handle_free(
	     &nsfdbexport_export_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free export handle.\n" );

		goto on_error;
	}
	if( log_handle_close(
	     log_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close log file: %" PRIs_SYSTEM ".\n",
		 log_filename );

		goto on_error;
	}
	if( log_handle_free(
	     &log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free log handle.\n" );

		goto on_error;
	}
	if( nsfdbexport_abort != 0 )
	{
		fprintf(
		 stdout,
		 "Export aborted.\n" );

		return( EXIT_FAILURE );
	}
	fprintf(
	 stdout,
	 "Export completed.\n" );

	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( log_handle != NULL )
	{
		log_handle_free(
		 &log_handle,
		 NULL );
	}
	if( nsfdbexport_export_handle != NULL )
	{
		export_handle_free(
		 &nsfdbexport_export_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

